protocol/op-supervisor-integration.md

```diff
@@ -0,0 +1,307 @@
+    2. The head-entry should be flushed to the DB, and tasks should be updated, once every L2 chain has been seen.
+       - TODO: multiple L2 blocks from the same L1 block. We need some way to report exhaustion of a L1 block.
+       - TODO: there may not be any L2 blocks derived from the L1 block. Can repeat parentL2 as "derived" in that case maybe?
+- If it does not match the task, then heck if it matches a past entry.
+  - If it does, silently accept the result. There may be multiple op-nodes running the same tasks.
+  - If it does not, then return an error. The op-node will need to adjust to the latest task.
+
+Note that the op-node can ignore reporting task-results,
+if the `CrossSafe()` is canonical to its view, and ahead of where the op-node is.
+
+`CrossSafe`: returns the current cross-safe block, at the time of `l1Number` L1 block number.
+The op-node should check if the return L1-block is canonical before accepting the result.
+
+
+### `TryFinalize`
+
+```go
+type RPC interface {
+    TryFinalize(chainID types.ChainID, finalizedL1 eth.BlockID) (finalizedL2 eth.BlockID, err error)
+}
+```
+
+- `TryFinalize`: this simply returns the latest L2 block of the specified `chainID` to finalize, which must be fully derived from the specified `finalizedL1` block.
+
+## Integration
+
+### op-node safety
+
+- Track `cross-unsafe` as new safety-level in the `EngineController`, similar to pseudo-levels like `backup-unsafe-reorg`, and `pending-safe`.
+- Modify the `engine` payload-processing step to not immediately mark a derived-from-L1 block as `safe`, but rather emit an event that indicates the `unsafe` block was derived and ready to promote. Don’t emit the pending-safe update immediately either.
+- Add new Interop event-deriver.
+    - Monitor the derive tasks of the super-visor with `NextDeriveTask(...)`
+      - Perform the tasks, report results with `OnDerived(...)`
+    - Poll `CrossSafe(...)` for cross-safe changes.
+    - Replace finalizer deriver, to simply take the L1 finality signal, forward it to the supervisor,
+      and get back a signal what to finalize on the L2 accordingly.
+      If the L1 finalization signal is too new, that’s ok, eventually we’ll have recent data,
+      or we can pass an older L1 block as input to the supervisor, to get an older L2 block to finalize.
+
+TODO:
+- We could add a mode to `op-node` where it ignores the derivation tasks,
+  and just takes `CrossSafe` to find a recent sync-target, and then trigger engine-sync with that.
+- We could expose a public `CrossSafe` endpoint, without the `NextDeriveTask`, `OnDerived`,
+  for low-resource users / replicas to have something trusted to maintain a safety view with.
+
+### op-geth safety
+
+During block-building:
+
+- After adding a tx to an in-progress block building job, inspect all the emitted new event-logs by said tx:
+    - For each executing message, query the op-supervisor `CheckMessage`, to verify if the message is sufficiently safe to include.
```

Comment:
It would be nice to have a spec for the API of the supervisor to help document it

Reply:
@ajsutton will help lift an API specs draft PR from this design-doc